### ISSUE TITLE
retrained on CS-campus3d

### ISSUE THREAD

**Issue by yandiyang** [csiro-robotics/HOTFormerLoc#5]
Hi Grifent,

Thank you for sharing the codes.

I retrained on CS-campus3d, with the default parameters and also NVIDIA N100 , but I got :

Avg. top 1% recall: 90.84   Avg. MRR: 87.54   Avg. recall @N:
[75.44853636 80.07554297 81.49197356 83.09726157 83.85269122 85.0802644
 86.40226629 86.77998111 86.96883853 87.62983947 88.101983   88.47969783
 88.76298395 88.85741265 88.95184136 89.23512748 89.32955619 89.42398489
 89.70727101 89.89612842 90.36827195 90.46270066 90.55712937 90.65155807
 90.84041549]
New best avg AR@1 at Epoch 265: 74.60 -> 75.45

However, the recall value on the paper are 80% and 95%. 

Also, still on CS-campus3d, I retrained without using the multi-stage training, I use the direct loss backward method (using your training_step codes). I set the batch_size:64, batch_split_size:0. val_batch_size: 64.I got :
Dataset: average
Avg. top 1% recall: 89.52   Avg. MRR: 80.93   Avg. recall @N:
[67.51652502 71.19924457 74.03210576 76.77053824 79.41454202 80.6421152
 81.30311615 81.96411709 82.71954674 83.66383381 84.13597734 84.70254958
 85.36355052 85.64683664 86.68555241 87.15769594 87.25212465 87.72426818
 87.72426818 88.38526912 88.66855524 89.14069877 89.42398489 89.5184136
 89.5184136 ]
New best avg AR@1 at Epoch 270: 67.33 -> 67.52

The recall 1 is a bit low.

Appreciate it if you can help me explain this. Thank you.

**Comment by Grifent** [comment]
Thanks for raising this issue.

Can I confirm that you trained for the full 300 epochs? If so, and this was the best AR@1 achieved, it is possibly due to the training instability on CS-Campus3D, as mentioned in [this issue](https://github.com/csiro-robotics/HOTFormerLoc/issues/4). In my experience, CS-Campus3D has unique training dynamics where the test set performance does not begin improving until >100 epochs, and tends to be quite unstable.

I have re-trained HOTFormerLoc on CS-Campus3D several times, and whilst it typically achieves similar performance to what is reported in the paper, it does occasionally get stuck in a local minima and fail to fully converge. 

See below for a comparison of the training loss and test set performance for two runs I trained on CS-Campus3D with identical configs. Training loss is nearly identical but the green run fails to fully converge on the test set (AR@1 ~= 76%). If this is the case, I would recommend re-training with the default config at least once or twice (ensure you use multi-stage training, as the large batch size of 2048 is essential). 

<img width="337" height="260" alt="Training Loss" src="https://github.com/user-attachments/assets/6311093b-c24d-465d-9465-6e0c57565cf9" />

<img width="337" height="260" alt="Image" src="https://github.com/user-attachments/assets/6e508385-cd77-40d9-a05a-2d94019e6183" />

It is unfortunately not possible to exactly reproduce training in PyTorch, even by setting random seeds, due to certain non-deterministic operations (see [here](https://docs.pytorch.org/docs/stable/notes/randomness.html)).

Let me know if you run into any further issues.

**Comment by Grifent** [comment]
I did try training without multi-stage in my initial experiments, and found a similar performance drop. However this is consistent on most datasets, as multi-stage backprop is required to train with large batch sizes (e.g. 2048), and the truncated smooth AP loss needs large batch sizes to function correctly, otherwise a triplet loss is better for smaller batch sizes (which can be set in the config if desired).

Functionally, the `training_step` and `multistaged_training_step` functions are mathematically identical for the same overall batch size. You should expect (roughly) the same performance using `training_step` with `batch_size=64` and `multistaged_training_step` with `batch_size=64` and `batch_split_size=32`, for example.

**Comment by yandiyang** [comment]
Thank you.
